Fix for UTF-8 BOM

anirudh2290 opened the pull request:
Ignore UTF-8 BOM character in CSV Parser if present.

@tqchen @piiswrong

anirudh2290 commented:
@cjolivier01 @reminisce @rahul003 @eric-haibin-lin  Can you guys please help review ?

cjolivier01 commented on src/data/text_parser.h:
```diff
@@ -48,32 +48,52 @@ class TextParserBase : public ParserImpl<IndexType> {
+    * \param begin the beginning position of buffer
+    * \return position of first endof line going backward
+    */
+  inline char *BackFindEndLine(char *bptr, char *begin) {
```
If it doesn't use any member variables, please make it static

cjolivier01 commented on src/data/text_parser.h:
```diff
@@ -48,32 +48,52 @@ class TextParserBase : public ParserImpl<IndexType> {
-    return begin;
+    if (count < 3)
+      *begin = *begin - count;
+    return;
```
return not needed here

haojin2 commented on src/data/text_parser.h:
```diff
@@ -48,32 +48,52 @@ class TextParserBase : public ParserImpl<IndexType> {
+    * \param end end of buffer
+    */
+  virtual void ParseBlock(char *begin, char *end,
+                           RowBlockContainer<IndexType> *out) = 0;
```
One nit, I guess this line should be aligned with the upper line?

eric-haibin-lin commented on src/data/text_parser.h:
```diff
@@ -48,32 +48,52 @@ class TextParserBase : public ParserImpl<IndexType> {
+    * \brief start from bptr, go backward and find first endof line
+    * \param bptr end position to go backward
+    * \param begin the beginning position of buffer
+    * \return position of first endof line going backward
```
clarify that it returns `begin` if not found in (begin, bptr]

anirudh2290 commented:
Thanks a lot @cjolivier01 , @haojin2 and @eric-haibin-lin for reviewing. I have made the necessary changes. Also changed to const char * and moved maxthread initialization outside the parallel region.

anirudh2290 commented:
@tqchen Can you please help review and merge this ?

tqchen commented:
wait until @cjolivier01 approve and i will merge this in

larroy commented on src/data/text_parser.h:
```diff
@@ -48,32 +44,51 @@ class TextParserBase : public ParserImpl<IndexType> {
+    * \return position of first endof line going backward, returns begin if not found
+    */
+  static inline const char *BackFindEndLine(const char *bptr, const char *begin) {
+     for (; bptr != begin; --bptr) {
```
we should assert that bptr >=begin

larroy commented on src/data/text_parser.h:
```diff
@@ -48,32 +44,51 @@ class TextParserBase : public ParserImpl<IndexType> {
-    for (; bptr != begin; --bptr) {
-      if (*bptr == '\n' || *bptr == '\r') return bptr;
+  static inline void IgnoreUTF8BOM(const char **begin, const char **end) {
+    int count = 0;
```
assert that end >= begin?
